core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -556,7 +569,7 @@ private void registerBuiltinQueryParsers() {
         registerQueryParser(new CommonTermsQueryParser(), CommonTermsQueryParser.QUERY_NAME_FIELD);
         registerQueryParser(new SpanMultiTermQueryParser(), SpanMultiTermQueryParser.QUERY_NAME_FIELD);
         QueryParser<FunctionScoreQueryBuilder> functionScoreParser = (QueryParseContext c) -> FunctionScoreQueryBuilder
-                .fromXContent((String name) -> functionScoreParsers.get(name), c);
+                .fromXContent(this::lookupScoreFunctionParser, c);
```

Comment:
I also struggle to understand this after looking for it for a while, maybe it's because I'm still not so fluent with this kind of syntax. The indirection makes it harder here. I'm not sure how adding another registry might help, but if it helps hining some of the complexity I'm all for it.